### PR TITLE
[SPARK-29606][SQL] Improve EliminateOuterJoin performance

### PR DESCRIPTION
**Pull request by wangyum**
### What changes were proposed in this pull request?

This PR try to improve `EliminateOuterJoin` performance via avoid generating too many constraints. For example:
```scala
import org.apache.spark.sql.catalyst.plans.logical.Project
spark.sql("CREATE TABLE IF NOT EXISTS spark_29606(a int, b int, c int) USING parquet")
spark.sql("SELECT a as a1, b as b1, c as c1, abc as abc1 FROM (SELECT a, b, c, a + b + c as abc FROM spark_29606) t")
  .queryExecution.analyzed.asInstanceOf[Project].validConstraints.toSeq.sortBy(_.toString).foreach(println)
```

**Before this PR**:
```
(((a#5 + b#6) + c#7) <=> abc#0)
(((a#5 + b#6) + c#7) <=> abc1#4)
(((a#5 + b#6) + c1#3) <=> abc#0)
(((a#5 + b#6) + c1#3) <=> abc1#4)
(((a#5 + b1#2) + c#7) <=> abc#0)
(((a#5 + b1#2) + c#7) <=> abc1#4)
(((a#5 + b1#2) + c1#3) <=> abc#0)
(((a#5 + b1#2) + c1#3) <=> abc1#4)
(((a1#1 + b#6) + c#7) <=> abc#0)
(((a1#1 + b#6) + c#7) <=> abc1#4)
(((a1#1 + b#6) + c1#3) <=> abc#0)
(((a1#1 + b#6) + c1#3) <=> abc1#4)
(((a1#1 + b1#2) + c#7) <=> abc#0)
(((a1#1 + b1#2) + c#7) <=> abc1#4)
(((a1#1 + b1#2) + c1#3) <=> abc#0)
(((a1#1 + b1#2) + c1#3) <=> abc1#4)
(a#5 <=> a1#1)
(abc#0 <=> abc1#4)
(b#6 <=> b1#2)
(c#7 <=> c1#3)
```

**After this PR**:
```
(((a#5 + b#6) + c#7) <=> abc#0)
(a#5 <=> a1#1)
(abc#0 <=> abc1#4)
(b#6 <=> b1#2)
(c#7 <=> c1#3)
```

### Why are the changes needed?
Improve `EliminateOuterJoin` performance.

**Before this PR**:
```
=== Metrics of Analyzer/Optimizer Rules ===
Total number of runs: 9323
Total time: 15.995000924 seconds

Rule                                                                                               Effective Time / Total Time                     Effective Runs / Total Runs

org.apache.spark.sql.catalyst.optimizer.EliminateOuterJoin                                         0 / 13359017999                                 0 / 4
org.apache.spark.sql.catalyst.analysis.Analyzer$ResolveRelations                                   990683801 / 991674120                           2 / 18
org.apache.spark.sql.catalyst.analysis.Analyzer$ResolveTables                                      0 / 443718064                                   0 / 18
org.apache.spark.sql.catalyst.analysis.DecimalPrecision                                            43087519 / 81709524                             2 / 18
org.apache.spark.sql.execution.datasources.PruneFileSourcePartitions                               63414650 / 63414650                             1 / 1
org.apache.spark.sql.catalyst.analysis.Analyzer$ResolveReferences                                  42587256 / 62760566                             5 / 18
...
```

**After this PR**:
```
=== Metrics of Analyzer/Optimizer Rules ===
Total number of runs: 9323
Total time: 3.03253427 seconds

Rule                                                                                               Effective Time / Total Time                     Effective Runs / Total Runs

org.apache.spark.sql.catalyst.analysis.Analyzer$ResolveRelations                                   1130336633 / 1131323257                         2 / 18
org.apache.spark.sql.catalyst.analysis.Analyzer$ResolveTables                                      0 / 448236638                                   0 / 18
org.apache.spark.sql.catalyst.optimizer.EliminateOuterJoin                                         0 / 107133411                                   0 / 4
org.apache.spark.sql.catalyst.analysis.DecimalPrecision                                            43965067 / 84085638                             2 / 18
org.apache.spark.sql.catalyst.analysis.Analyzer$ResolveReferences                                  44448673 / 66690250                             5 / 18
org.apache.spark.sql.execution.datasources.PruneFileSourcePartitions                               59631169 / 59631169                             1 / 1
...
```

### Does this PR introduce any user-facing change?

No.

### How was this patch tested?

Unit test.

### CODE REVIEW

**Comment by maropu** [comment]
still WIP?

**Comment by wangyum** [comment]
Thank you @maropu Actually, I am not very confident about this change. Is it make sense for you?

**Comment by maropu** [comment]
Ur, I see. ok, I've not digged into this, so I'll check later.

**Comment by wangyum** [comment]
cc @cloud-fan @viirya

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/LogicalPlan.scala [hunk 1]
```diff
@@ -176,10 +176,16 @@ abstract class UnaryNode extends LogicalPlan {
         allConstraints += EqualNullSafe(a.toAttribute, l)
       case a @ Alias(e, _) =>
         // For every alias in `projectList`, replace the reference in constraints by its attribute.
-        allConstraints ++= allConstraints.map(_ transform {
-          case expr: Expression if expr.semanticEquals(e) =>
-            a.toAttribute
-        })
+        allConstraints ++= allConstraints.map {
+          case e @ EqualNullSafe(l, _: AttributeReference)
+            if !l.isInstanceOf[AttributeReference] => e
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/LogicalPlan.scala]: what's the rationale behind this change?

Comment (wangyum) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/LogicalPlan.scala]: For example:
```scala
import org.apache.spark.sql.catalyst.plans.logical.Project
spark.sql("CREATE TABLE IF NOT EXISTS spark_29606(a int, b int, c int) USING parquet")
spark.sql("SELECT a as a1, b as b1, c as c1, abc as abc1 FROM (SELECT a, b, c, a + b + c as abc FROM spark_29606) t")
  .queryExecution.analyzed.asInstanceOf[Project].validConstraints.toSeq.sortBy(_.toString).foreach(println)
```
`child.constraints` has a  constraint: `(((a#5 + b#6) + c#7) <=> abc#0)`:


image


**Before this PR**.  We will replace the reference in constraints by its attribute. It will generate a lot of constraints. It seems that these constraints are useless:
image


**After this PR**: We avoid generating these useless constraints:
image

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/LogicalPlan.scala]: This is just an example. Why can we ignore `EqualNullSafe` between attr and non-attr? Are they guaranteed to produce useless constraints?

Comment (wangyum) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/LogicalPlan.scala]: `validConstraints` is used here:
https://github.com/apache/spark/blob/15de6d05008cf8054fe9c2246ee80f607f9f33e0/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/QueryPlanConstraints.scala#L30-L43

1. These generated constraint are not the `subsetOf(outputSet)`.
2. These generated constraint will not affect `inferAdditionalConstraints` and `constructIsNotNullConstraints `.

**Comment by wangyum** [comment]
Metrics of Analyzer/Optimizer Rules for `TPCDSQuerySuite`.
Before this PR:
```
05:43:31.429 WARN org.apache.spark.sql.TPCDSQuerySuite: 
=== Metrics of Analyzer/Optimizer Rules ===
Total number of runs: 224786
Total time: 65.253508851 seconds
...
```

After this PR:
```
05:38:55.596 WARN org.apache.spark.sql.TPCDSQuerySuite: 
=== Metrics of Analyzer/Optimizer Rules ===
Total number of runs: 224786
Total time: 56.81012364 seconds
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/LogicalPlan.scala [hunk 2]
```diff
@@ -176,10 +176,13 @@ abstract class UnaryNode extends LogicalPlan {
         allConstraints += EqualNullSafe(a.toAttribute, l)
       case a @ Alias(e, _) =>
         // For every alias in `projectList`, replace the reference in constraints by its attribute.
-        allConstraints ++= allConstraints.map(_ transform {
-          case expr: Expression if expr.semanticEquals(e) =>
-            a.toAttribute
-        })
+        allConstraints ++= allConstraints.map {
+          case e @ EqualNullSafe(l, _: AttributeReference) if l.references.size > 1 => e
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/LogicalPlan.scala:180]: I feel its a bit difficult to understand this pattern-matching at a glance, could you leave comments about what this means? Probably, you wanna skip [the pattern below](https://github.com/apache/spark/pull/26257/files#diff-27c76f96a7b2733ecfd6f46a1716e153R186) for performance?

**Comment by maropu** [comment]
retest this please
